### ISSUE TITLE
Dynamic threading

### ISSUE THREAD

**Issue by zarajafari** [Arash-codedev/openGA#13]
Hi Arash and Luc,

I have been using your library to empirically calibrate parameters of a simulation model. I hold the value for dynamic_threading true. When I run the program on my quad core Mac, all the cores (8 threads) are being used at their max capacity as I control them via Monitor activity. The project is computationally expensive when I run the simulation for the experiments at larger size. Therefore, I run the code on a server at my office which is a AMD EPYC 32-core processor. I expected this to improve the performance in terms of speed, but it did not. When I monitored the CPU activity via htop, I noticed it does not efficiently maximise the usage of each thread as you can see in the screenshot below. And even once I limit the N_thread value to 8 or 4, it does not maximise the activity of any 4 or 8 threads, which is something I did not expect. 

I am not sure if you have any idea why this might happen? I would appreciate it if you could share any tip or guess that might help me solve the issue. 

Many thanks,
Zahra

<img width="692" alt="Screenshot 2020-12-21 at 12 22 53" src="https://user-images.githubusercontent.com/53267747/102778351-54ab6a80-438a-11eb-83da-b26da766f05d.png">

**Comment by Arash-codedev** [comment]
Hi Zahra,
Thank you very much for using openGA. I suspect a few assumptions.
First, if a single solution (not the whole population) is solved quickly, then you probably do not need dynamic threading. Static threading will be faster.
Also, please set "ga_obj.idle_delay_us" to zero. Otherwise the switching between threads will have some delays. This will be useful when each of your single solution takes 40 seconds to calculate to prevent overloading a CPU unnecessarily. However, when your single solution can be evaluated quickly, that will create more problem than benefit.
Please update me about the impact of the outcome.

Best regards,
Arash

**Comment by zarajafari** [comment]
Hi Arash,

Thanks a lot for your response. 

I evaluated the runtime for a single solution, which is on average about 1 minute and 15 seconds (this can grow larger once I increase the scale of my program). Given this, would you still recommend static threading? 

At the moment, "idle_delay_us"  is set to 1, which is something I did not notice before. This means it switches between threads so quickly, do you think this might be just too quick regarding the time each single solution requires?

I attempted the static threading and I monitored the threads. It seems like 2-3 (constantly varying) threads are used at their partial capacities as you can see in the screenshot attached below.

Many thanks,
Zahra

![IMAGE 2021-01-04 14:16:34](https://user-images.githubusercontent.com/53267747/103544102-6ff09c80-4e97-11eb-879c-c57fb7055953.jpg)

**Comment by Arash-codedev** [comment]
Hi Zahra,
Sorry for responding you late. However, I check my email frequently.
In C++, there is a function `std::thread::hardware_concurrency()` which gives the number of available cores. OpenGA is using this to obtain the number of the cores. You can assign the `N_threads` field manually as much as you like. Would you please create an instance of GA object and print the value of `ga_obj.N_threads`? It shows how many threads it is creating. Please also check you are not printing too much output in your console as it delays your application more than what most of people guess.
To get more information from OSX you may use this command:
```
sudo dtruss -ap pid
```
Please also make sure your threads do not block each other by calling common OS resources such as disk or network or memory or even console print. 
Please also make sure when compile you use `-O3` option as well as `-pthread` for your gcc/clang compiler.
Please also make sure you are not limited by memory or other resources.
If none of the above attempts help you, then please grab one of the existing examples and increase it population high or logic to hard so to challenge your CPU. If you still utilize all the cores, I appreciate you ask it as a MWE (minimal working example) on stackoverflow.com. As my laptop has no more than 4 cores, I cannot replicate this problem on my own laptop. Solving your problem will possibly help others too. Thanks.

**Comment by zarajafari** [comment]
Thank you very much for all the responses you provided.

I did all the steps you mentioned and I think I figured out where the issue lies. I guess the problem arises partly because the model I aim to optimize was computationally expensive; it was not parallelized and optimized well within its own frame. More importantly, there was a memory problem that extremely affected the multithreading in OpenGA, the memory is a big problem when it comes to creating a pool of solutions. I am going to use SQLite to optimize the performance as my code involves loops with multiple access to memory. Anyways, thanks a lot for all the support, I learned a lot!

Best, 
Zahra
